File: benchmarking/benchmarks/Utility-ML-Functions/javascript/pooling.ts
/**
 * @license
 * Copyright 2017 Google Inc. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * =============================================================================
 */

import * as dl from 'deeplearn';

import {BenchmarkTest} from './benchmark';
import * as benchmark_util from './utilities';

export interface PoolBenchmarkParams {
  depth: number;
  fieldSize: number;
  stride: number;
}

function getPoolingOp(option: string): (x: dl.Tensor3D, filterSize: [number, number]|number,
                                        strides: [number, number]|number) => dl.Tensor3D {
    switch (option) {
        case 'max':
            return (x: dl.Tensor3D, filterSize: [number, number]|number, strides: [number, number]|number) => {
                return x.maxPool(filterSize, strides, 'same');
            };
        case 'avg':
            return (x: dl.Tensor3D, filterSize: [number, number]|number, strides: [number, number]|number) => {
                return x.avgPool(filterSize, strides, 'same');
            };
        default:
            throw new Error(`Not found such ops: ${option}`);
    }
}

export class PoolingCPU implements BenchmarkTest {
    run(size: number, option: string, params: PoolBenchmarkParams): Promise<number> {
        dl.setBackend('cpu');

        const outputDepth = params.depth;
        const xShape: [number, number, number] = [size, size, outputDepth];
        const fieldSize = params.fieldSize;
        const stride = params.stride;
        const op = getPoolingOp(option);

        const x: dl.Tensor3D = dl.randomUniform(xShape, -1, 1);

        const start = performance.now();

        op(x, fieldSize, stride);

        const avgTime = (performance.now() - start);
        return Promise.resolve(avgTime)
    }
}

export class PoolingGPU implements BenchmarkTest {
    async run(size: number, option: string, params: PoolBenchmarkParams): Promise<number> {
        dl.setBackend('webgl');

        const outputDepth = params.depth;
        const xShape: [number, number, number] = [size, size, outputDepth];
        const fieldSize = params.fieldSize;
        const stride = params.stride;
        const x: dl.Tensor3D = dl.randomUniform(xShape, -1, 1);
        const op = getPoolingOp(option);

        const benchmark = () => op(x, fieldSize, stride);
        const time = await benchmark_util.warmupAndBenchmarkGPU(benchmark);

        x.dispose();

        return time;
    }
}

export async function run(size, type, params, backend){
    let benchmark: BenchmarkTest = (backend === 'gpu') ? new PoolingGPU() : new PoolingCPU();
    return await benchmark.run(size, type, params)
}

File: utility-benchmarks/python/pooling.py
from __future__ import absolute_import
from __future__ import division
from __future__ import print_function

import tensorflow as tf
import utilities as util
import numpy as np
import time

config = tf.ConfigProto()
config.gpu_options.allow_growth = True


def get_pooling_op(x, pool_size, strides, option):
    if option == 'max':
        return tf.layers.max_pooling1d(x, pool_size, strides, 'same')
    else:
        assert option == 'avg', 'Pooling option not found: ' + option
        return tf.layers.average_pooling1d(x, pool_size, strides, 'same')


def run(size, option, params, backend):
    if backend == 'gpu' and util.can_use_gpu():
        doWarmup = True
    else:
        assert backend == 'cpu'
        doWarmup = False

    output_depth = params['depth']
    x_shape = np.asarray([size, size, output_depth], dtype=np.int32)
    field_size = params['field_size']
    stride = params['stride']
    x = tf.random_uniform(x_shape, -1, 1)
    pool = get_pooling_op(x, field_size, stride, option)

    backend = util.get_backend(backend)
    with tf.device(backend):
        sess = tf.Session()

        if doWarmup:
            sess.run(pool)

        start = time.time()
        sess.run(pool)
        end = time.time()
        runtime = end - start
        print(runtime)


if __name__ == "__main__":
    run(100, 'max', {'depth': 10, 'field_size': 2, 'stride': 2}, 'cpu')